File: day10/RotateList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        // Check for base case
        if (k == 0) return head;
        // Check if rotation is actually required
        if (head == null || head.next == null) return head;
        
        // Find the node before the last node,
    
        ListNode beforeLast = head;
        //after checking those conditions above we have minimum lenth of list is 2
        int length = 2;
        while (beforeLast.next.next != null) {
            beforeLast = beforeLast.next;
            length += 1;
        }

        // Do one rotation
        beforeLast.next.next = head;
        head = beforeLast.next;
        beforeLast.next = null;

        // rotate again for next k
        return rotateRight(head, (k - 1) % length);
    }
}

           

File: day10/Rotate_list.cpp
/*
*********auther- Anand keshari****************

problem link- https://leetcode.com/problems/rotate-list/

***********Time Complexity- O(n)************************
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head==NULL)
            return head;
        ListNode* ptr1= new ListNode;
        ListNode* ptr2= new ListNode;
        ptr1=head;
        ptr2=head;
        int cnt=1;
        while(ptr2->next!=NULL) 
        {
            cnt++;
            ptr2=ptr2->next;
        }
        k=k%cnt;
        k=cnt-k-1;
       // if(k==0)
         //   return head;
        while(k>0)
        {
            ptr1=ptr1->next;
            k--;
        }
        ptr2->next=head;
        head=ptr1->next;
        ptr1->next=NULL;
        return head;
    }
};
